lib/data_migrator/offence_unique_code_migrator.rb

```diff
@@ -2,41 +2,50 @@
 
 module DataMigrator
   class OffenceUniqueCodeMigrator
-    attr_reader :offences
+    attr_reader :offences, :offence_set
 
-    def initialize
-      @offences = Offence.unscoped.order(:description)
-      create_offence_set
-    end
+    class InappropriateRelation < StandardError; end
 
-    def migrate!
-      puts '-- updating offences.unique_code data'
-      @offence_set.each do |code, details|
-        sql = "UPDATE offences SET unique_code = \'#{code}\' WHERE id = #{details[:id]}"
-        @offences.connection.execute sql
-      end
+    def initialize(relation: nil)
+      raise InappropriateRelation, "Inappropriate relation given: expected #{Offence}, got #{relation.klass}" unless relation.nil? || relation.klass.eql?(Offence)
+      @offences = relation || Offence.unscoped.order(description: :asc)
+      create_offence_set
     end
 
     def create_offence_set
       @offence_set = {}
       offences.each_with_object(@offence_set) do |offence, set|
-        set[unique_code(offence)] = { id: offence.id, description: offence.description }
+        set[unique_code(offence)] = { id: offence.id, description: offence.description, unique_code: offence.unique_code, contrary: offence.contrary, band: offence.offence_band&.description, class_letter: offence.offence_class&.class_letter }
         set
       end
     end
 
-    def test
-      @offence_set.each do |code, details|
-        ap "UPDATE offences SET unique_code = #{code} WHERE id = #{details[:id]} \# for #{details[:description]}"
+    def migrate!
+      puts '-- clearing existing offences.unique_code data'
```

Comment:
**[nit pick/not a blocker]** 💭 Consider using the `Rails.logger`. You can always configure the logger to be set to STDOUT which will have the same effect as this, or redirect the logs to stdout if it's not by default.